File: ios/AtlasCore/Sources/AtlasCore/Keys/Derivation.swift
import Foundation

/// Session key derivation, forward-secret ratchet, contexts and tokens
/// (§2.2, §2.3). Mirrors `backend/atlas/keys/derivation.py` + `tokens.py`.
public enum KeyError: Error { case destroyed }

/// RAM-only session key (§2.2). `destroy()` zeroises it — the primary
/// containment mechanism. Reference type so a wipe is observed by all holders.
///
/// ROLE SEPARATION (§2.3). This is a ROOT, not a working key. Nothing outside
/// this class should consume the raw bytes: every consumer takes a purpose-scoped
/// leaf via `contextKey()` (`storage`, `recognition`, `tunnel`, `chain`,
/// `continuity`). The epoch chain and the continuity chain tick on different
/// clocks and are exposed through different surfaces; handing the same 32 bytes
/// to both turns one compromise into three.
///
/// KEY LIFETIME. The buffer is EXPLICITLY allocated rather than held in a Swift
/// `Array`/`Data`: those are copy-on-write and the runtime may reallocate them,
/// leaving stale plaintext at the old address that no wipe can reach. One
/// allocation, one address, wiped in place, freed on `deinit`.
///
/// The wipe goes through `memset_s` where available. A plain zeroing loop over
/// memory that is never read again is a dead store, and an optimising compiler
/// is entitled to delete it — which is the reason `memset_s`/`explicit_bzero`
/// exist at all.
///
/// Prefer `withKey { }` (borrow, no copy) over `.key` (escaping copy) anywhere
/// the caller does not need to RETAIN the bytes past the call: `destroy()` can
/// only reach the buffer this object owns, so containment is exactly as strong
/// as the copies that were never made.
public final class SessionKey {
    public let drandRound: Data
    private let buf: UnsafeMutableRawBufferPointer
    private(set) public var alive = true

    init(drandRound: Data, key: Data) {
        self.drandRound = drandRound
        // Explicit allocation: fixed address, no COW, no silent realloc.
        self.buf = UnsafeMutableRawBufferPointer.allocate(
            byteCount: key.count, alignment: MemoryLayout<UInt8>.alignment)
        key.withUnsafeBytes { self.buf.copyMemory(from: $0) }
    }

    /// Borrow the key material without minting a copy the wipe cannot reach.
    /// The pointer is valid ONLY for the duration of `body` — do not escape it.
    public func withKey<T>(_ body: (UnsafeRawBufferPointer) throws -> T) throws -> T {
        guard alive else { throw KeyError.destroyed }
        return try body(UnsafeRawBufferPointer(buf))
    }

    /// ESCAPING COPY — the returned `Data` outlives `destroy()`. Kept for the
    /// callers that legitimately RETAIN key bytes (the ratchet chain feeds
    /// `prevKey` forward, see `Session/Device.swift`). Everything else should
    /// use `withKey`.
    public var key: Data {
        get throws {
            guard alive else { throw KeyError.destroyed }
            return Data(buf)
        }
    }

    /// Derive a purpose-scoped LEAF. Every consumer of a session key takes one of
    /// these; nothing outside this class consumes the raw root. HKDF is one-way,
    /// so a leaked leaf yields neither the root nor a sibling leaf.
    public func contextKey(_ context: String) throws -> Data {
        let info: Data
        switch context {
        case "storage": info = Params.contextStorage
        case "recognition": info = Params.contextRecognition
        case "tunnel": info = Params.contextTunnel
        case "chain": info = Params.contextChain
        case "continuity": info = Params.contextContinuity
        default: fatalError("unknown context \(context)")
        }
        return try withKey { raw in
            // Transient — hkdfCombine takes [Data]; wipe it before it is dropped.
            var ikm = Data(raw)
            defer { ikm.withUnsafeMutableBytes { SessionKey.wipe($0) } }
            return Primitives.hkdfCombine([ikm], info: info, length: 32)
        }
    }

    /// Zeroise the key (the primary containment mechanism, §2.2). Idempotent.
    public func destroy() {
        SessionKey.wipe(buf)
        alive = false
    }

    deinit {
        // A key dropped without an explicit destroy() must not leave plaintext
        // behind for the allocator to hand out again.
        SessionKey.wipe(buf)
        buf.deallocate()
    }

    static func wipe(_ b: UnsafeMutableRawBufferPointer) { secureWipe(b) }
}

/// Wipe that the optimiser is not permitted to elide. A zeroing loop over memory
/// never read again is a dead store the compiler may delete — which is precisely
/// why `memset_s`/`explicit_bzero` exist.
func secureWipe(_ b: UnsafeMutableRawBufferPointer) {
    guard let base = b.baseAddress, b.count > 0 else { return }
    #if canImport(Darwin)
    _ = memset_s(base, b.count, 0, b.count)
    #else
    // No memset_s: write through a runtime-obtained pointer (the optimiser
    // cannot prove the store dead) and keep the buffer alive across it.
    let p = base.assumingMemoryBound(to: UInt8.self)
    for i in 0..<b.count { p.advanced(by: i).pointee = 0 }
    withExtendedLifetime(b) {}
    #endif
}

/// A wipeable buffer for key material that must be HELD ACROSS CALLS.
///
/// `Data` cannot be wiped: reassigning (`x = Data(repeating: 0, count: 32)`)
/// drops the old copy-on-write allocation intact for the allocator to hand out
/// again. Anything that RETAINS key material on a long-lived object — the
/// ratchet's prev-key, the continuity chain — must live here instead, so a
/// device seized right after a liveness break holds no usable material
/// (§2.2, §5.3). Mirrors `SecretBytes` in `backend/atlas/keys/derivation.py`.
///
/// `setFrom()` overwrites IN PLACE; the buffer's address never changes, so
/// there is never a second copy to chase.
public final class SecretBytes {
    private let buf: UnsafeMutableRawBufferPointer

    public init(count: Int = 32) {
        buf = UnsafeMutableRawBufferPointer.allocate(
            byteCount: count, alignment: MemoryLayout<UInt8>.alignment)
        secureWipe(buf)
    }

    /// Borrow the material without minting a copy the wipe cannot reach.
    /// Valid ONLY for the duration of `body` — do not escape it.
    public func withBytes<T>(_ body: (UnsafeRawBufferPointer) throws -> T) rethrows -> T {
        try body(UnsafeRawBufferPointer(buf))
    }

    /// Overwrite in place from any raw buffer of the same length.
    public func setFrom(_ src: UnsafeRawBufferPointer) {
        precondition(src.count == buf.count, "length mismatch")
        buf.copyMemory(from: src)
    }

    public func setFrom(_ src: Data) {
        src.withUnsafeBytes { setFrom($0) }
    }

    /// Zeroise in place. Idempotent.
    public func wipe() { secureWipe(buf) }

    deinit {
        secureWipe(buf)
        buf.deallocate()
    }
}

public enum Derivation {
    /// SessKey = HKDF(PoLE_value, LK, epoch_key, prev_key, ctx) (§2.3).
    /// `poleValue` is a physio-timed clean QRNG value (the ring's live signal
    /// timed the firing; the value is clean QRNG). No continuity flag, no raw
    /// physiology, no drand. Input list order preserved for cross-impl parity.
    public static func sessionKeyDecoupled(lk: Data, epochKey: Data, poleValue: Data,
                                           prevKey: Data, contextSeparator: Data,
                                           drandRound: Data) -> SessionKey {
        let m = Primitives.hkdfCombine([lk, epochKey, poleValue, prevKey, contextSeparator],
                                       info: Params.labelSession, length: 32)
        return SessionKey(drandRound: drandRound, key: m)
    }

    /// Claimed embodiment (Math Spec §A); built for parity, not the default.
    public static func sessionKeyCoupled(tsk: Data, devKey: Data, poleState: Data,
                                         beacon: Data, drandRound: Data) -> SessionKey {
        let m = Primitives.hkdfCombine([tsk, devKey, poleState, beacon],
                                       info: Params.labelSession, length: 32)
        return SessionKey(drandRound: drandRound, key: m)
    }

    /// Forward-secret ratchet (§2.2):
    /// K[t+1] = HKDF( K[t] || H(entropy_t) || beacon_t || drand_round ).
    public static func ratchet(_ prevKey: Data, entropyT: Data, beaconT: Data, drandRound: Data) -> Data {
        Primitives.hkdfCombine([prevKey, Primitives.H(entropyT), beaconT, drandRound],
                               info: Params.labelRatchet, length: 32)
    }
}

File: backend/atlas/keys/derivation.py
"""Session key derivation, forward-secret ratchet, contexts and tokens (§2.2, §2.3).

Locked Model §2.3 — the ONE principle: value = QRNG (clean); timing/liveness
times draws and gates operations but NEVER enters a value or KDF.

  SessKey = HKDF( PoLE_value, LK, epoch_key, prev_key, context_separator )

  * PoLE_value      — a physiologically-TIMED QRNG value: the enrolled ring's
                      live signal times WHEN the device QRNG fires; the fired
                      value is clean QRNG (raw physiology never enters it). This
                      replaces the earlier un-timed `local_qrng_draw`.
  * LK, epoch_key   — clean QRNG values; present ONLY because continuity gated
                      their unwrap (epoch key unwraps LK; see session/presence).
  * NO continuity_flag (continuity is upstream — it gates the unwrap, it is not a
    KDF ingredient), NO raw physiology, NO drand.

(The `coupled` variant below is a legacy alternative claim, not the default.)

Forward-secret ratchet (§2.2):
  K[t+1] = HKDF( K[t] || H(entropy_t) || beacon_t || drand_round )

SessKey is RAM-only and must be destroyed on liveness break / logout /
attestation failure / epoch boundary (§2.2). `SessionKey.destroy()` zeroises it;
the containment tests assert a destroyed key cannot decrypt.
"""

from __future__ import annotations

from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Iterator

from ..crypto.primitives import H, hkdf_combine
from ..params import (
    CONTEXT_RECOGNITION,
    CONTEXT_STORAGE,
    CONTEXT_TUNNEL,
    LABEL_RATCHET,
    LABEL_SESSION,
)

_CONTEXTS = {
    "storage": CONTEXT_STORAGE,
    "recognition": CONTEXT_RECOGNITION,
    "tunnel": CONTEXT_TUNNEL,
}


class KeyDestroyedError(RuntimeError):
    """Raised when a destroyed (RAM-wiped) session key is used."""


@dataclass
class SessionKey:
    """RAM-only session key material (§2.2). Never written to storage.

    KEY LIFETIME (§2.2 containment). `destroy()` zeroises the ONE buffer this
    object owns. It cannot reach a copy someone else already took, so the
    containment guarantee is only ever as strong as the copies that were never
    made. Prefer `borrow()` (no copy) over `.key` (escaping copy) everywhere a
    caller does not need to RETAIN the bytes past the call.

    HONEST BOUNDARY — what CPython cannot guarantee (stated exactly, cf. the
    Swift `SessionKey`, which can and does do better):
      * `bytes` is immutable, so any copy handed out — via `.key` — is
        unreachable by the wipe and lives until the GC frees it. Freed memory is
        not zeroised.
      * The interpreter may itself copy the buffer (realloc, refcount churn),
        leaving stale plaintext behind us.
      * Nothing is pinned, so the page may reach swap or a core dump.
    Treat Python-side zeroisation as BEST EFFORT for the reference
    implementation. The forensic-resistance claim ("RAM holds no key once
    liveness breaks") rests on the device implementations, where the buffer is
    explicitly allocated, wiped through `memset_s`, and freed on `deinit`.
    """

    drand_round: bytes
    _key: bytearray = field(repr=False)
    _alive: bool = True

    @contextmanager
    def borrow(self) -> Iterator[memoryview]:
        """Borrow the key material WITHOUT minting a copy the wipe cannot reach.

        Yields a read-only `memoryview` over the live buffer — valid only inside
        the `with` block. Do NOT stash it; `destroy()` zeroises underneath it by
        design (item assignment stays legal while a view is exported)."""
        if not self._alive:
            raise KeyDestroyedError("session key was destroyed (liveness/epoch)")
        view = memoryview(self._key).toreadonly()
        try:
            yield view
        finally:
            view.release()

    @property
    def key(self) -> bytes:
        """ESCAPING COPY — the returned `bytes` outlives `destroy()`.

        Kept for the callers that legitimately RETAIN key bytes (the ratchet
        chain feeds `prev_key` forward, see `session/device.py`). Everything
        else should use `borrow()`."""
        if not self._alive:
            raise KeyDestroyedError("session key was destroyed (liveness/epoch)")
        return bytes(self._key)

    def context_key(self, context: str) -> bytes:
        """Derive a purpose-scoped key (§2.3 HKDF info label per purpose)."""
        if context not in _CONTEXTS:
            raise ValueError(f"unknown context {context!r}")
        with self.borrow() as k:                      # no escaping copy of the root key
            return hkdf_combine([k], info=_CONTEXTS[context], length=32)

    def destroy(self) -> None:
        """Zeroise the key (the primary containment mechanism, §2.2). Idempotent."""
        for i in range(len(self._key)):
            self._key[i] = 0
        self._alive = False

    def __del__(self) -> None:
        # A key that is dropped without an explicit destroy() must not leave
        # plaintext behind for the allocator to hand out. Interpreter shutdown
        # can null out globals under us, so this is deliberately total.
        try:
            self.destroy()
        except Exception:
            pass

    @property
    def alive(self) -> bool:
        return self._alive


def derive_session_key_decoupled(
    *,
    lk: bytes,
    epoch_key: bytes,
    pole_value: bytes,
    prev_key: bytes,
    context_separator: bytes,
    drand_round: bytes,
) -> SessionKey:
    """Session key = HKDF(PoLE_value, LK, epoch_key, prev_key, ctx) (§2.3).

    `pole_value` is the physiologically-TIMED QRNG value (clean QRNG whose firing
    was timed by the ring's live signal). Continuity is NOT an input here — it is
    upstream, gating whether LK/epoch_key could be unwrapped at all. No raw
    physiology, no drand. (Input list order is preserved for cross-impl parity.)"""
    material = hkdf_combine(
        [lk, epoch_key, pole_value, prev_key, context_separator],
        info=LABEL_SESSION,
        length=32,
    )
    return SessionKey(drand_round=drand_round, _key=bytearray(material))


def derive_session_key_coupled(
    *, tsk: bytes, dev_key: bytes, pole_state: bytes, beacon: bytes, drand_round: bytes
) -> SessionKey:
    """REFERENCE ONLY — NOT a live code path. Do NOT wire into session derivation.

    The Math Spec §A "coupled" embodiment folds the PoLE *state* into the KDF. We
    DELIBERATELY DO NOT use it: mixing low-entropy, estimable, non-reproducible
    liveness/timing data into key material can only weaken the key (shrinks the
    keyspace, and a measurement of your biology would then help derive the key). The
    presence BINDING it aims for is already provided — cleanly — by GATING the unwrap
    of LK/epoch_key upstream (see `derive_session_key_decoupled` + session/presence).

    DECISION (stands): value = QRNG (clean); timing/liveness TIMES the QRNG firing and
    GATES operations, but NEVER enters a value/KDF. This function exists only so its
    behavior is pinned (test_coupled_epoch_gaps.py); the live path is the decoupled one.
    """
    material = hkdf_combine(
        [tsk, dev_key, pole_state, beacon], info=LABEL_SESSION, length=32
    )
    return SessionKey(drand_round=drand_round, _key=bytearray(material))


def ratchet(prev_key: bytes, *, entropy_t: bytes, beacon_t: bytes, drand_round: bytes) -> bytes:
    """Forward-secret ratchet step (§2.2).

    K[t+1] = HKDF( K[t] || H(entropy_t) || beacon_t || drand_round ).
    One-way: K[t] cannot be recovered from K[t+1] (a captured later key cannot
    read earlier messages, and vice-versa).
    """
    return hkdf_combine(
        [prev_key, H(entropy_t), beacon_t, drand_round], info=LABEL_RATCHET, length=32
    )
